Regular Expressions/RegExp (Basic).sublime-syntax

```diff
@@ -0,0 +1,280 @@
+      captures:
+        1: storage.modifier.mode.regexp
+        2: keyword.control.group.regexp
+      set: [base-group-extended, maybe-unexpected-quantifiers]
+    # Groups with 'x' mode
+    - match: '\?[ims]*x[ixms]*(?:-[ims]+)?:'
+      scope: storage.modifier.mode.regexp
+      set: [group-body-extended, maybe-unexpected-quantifiers]
+    # Other modifiers
+    - match: '\?(?:[ixms]*-)?[ixms]+:'
+      scope: storage.modifier.mode.regexp
+      set: [group-body, maybe-unexpected-quantifiers]
+    - include: group-start-common
+    - match: ''
+      set: [group-body, maybe-unexpected-quantifiers]
+
+  group-start-extended:
+    - meta_scope: meta.group.extended.regexp
+    - match: \?(<[=!]|[>=:!])
+      scope: constant.other.assertion.regexp
+      set: [group-body-extended, maybe-unexpected-quantifiers]
+    # Deactivates 'x' mode
+    - match: (\?[ims]*-[ims]*x[imxs]*)(\))
+      captures:
+        1: storage.modifier.mode.regexp
+        2: keyword.control.group.regexp
+      set: [base-group, maybe-unexpected-quantifiers]
+    # Groups without 'x' mode
+    - match: '\?[ims]*-[ims]*x[imxs]*:'
+      scope: storage.modifier.mode.regexp
+      set: [group-body, maybe-unexpected-quantifiers]
+    # Other modifiers
+    - match: '\?(?:[ixms]*-)?[ims]+:'
+      scope: storage.modifier.mode.regexp
+      set: [group-body-extended, maybe-unexpected-quantifiers]
+    - include: group-start-common
+    - match: ''
+      set: [group-body-extended, maybe-unexpected-quantifiers]
+
+  group-start-common:
+    # Other modifiers (must come after other modifier matches)
+    - match: (\?(?:[ixms]*-)?[ixms]+)(\))
+      captures:
+        1: storage.modifier.mode.regexp
+        2: keyword.control.group.regexp
+      pop: 1
+
+  group-body:
+    - meta_content_scope: meta.group.regexp
+    - match: \)
+      scope: meta.group.regexp keyword.control.group.regexp
+      pop: 1
+    - include: base-literal
+
+  group-body-extended:
+    - meta_content_scope: meta.group.extended.regexp
+    - match: \)
+      scope: meta.group.extended.regexp keyword.control.group.regexp
+      pop: 1
+    - include: base-literal-extended
+
+  charsets:
+    - match: (\[\^?)\]?
+      captures:
+        1: keyword.control.set.regexp
+      push: charset-body
+
+  charset-body:
+    - meta_scope: meta.set.regexp
+    - match: \]
+      scope: keyword.control.set.regexp
+      pop: 1
+    - include: charset-ranges
+    - include: character-classes
+    - include: escaped-chars
+
+  charset-ranges:
+    - match: (?=({{known_char_escape}}|{{char_escape}}|(?!\\-)[^\]])-({{known_char_escape}}|{{char_escape}}|[^\]]))
+      push: [charset-range-end, charset-range-begin]
+
+  charset-range-begin:
+    - match: '-'
+      scope: punctuation.separator.sequence.regexp-common
+      pop: 1
+    - include: escaped-chars
+
+  charset-range-end:
+    - meta_scope: constant.other.range.regexp
+    - include: escaped-char
+    - match: .
+      pop: 1
```

Comment:
With a look at PHP and Python, those use a single pattern with capture groups to match charset-ranges.

Do you think it makes sense to do so, too?

I've created a prototype for it and it performs slightly faster than this lookahead construct. It would make the popping `escaped-char` context obsolete, but result in some more scope assignments due to capturing groups.

Thoughts?

Reply:
No strong opinions here, just a few questions to help us decide which approach to take:

- is (slightly) better performance our main goal?
- is easier maintenance via less repeated scopes (and more re-used contexts) our main goal?
- if it is a mix between the two, does the performance difference outweigh the "duplication"?
- which is easier to understand - a complex lookahead which is then broken down into match patterns / context includes, or a consuming regex with many capture groups?

I'd personally even want to consider using a branch point instead of a lookahead or a single regex, to make it easier to follow what is happening, prevent duplication of scope assignment in various places etc. But that may make performance even worse than the current solution, not sure.

Reply:
That's correct. I have the same questions in mind.

Additionally I think about the theoretical fact of what' happens happens if the syntax is included via `with_prototype`.

My prototype lookes as follows:

```suggestion
  charset-ranges:
    - match: |-
        (?x)
        (?: ({{known_char_escape}}) | ({{invalid_char_escape}}) | ({{char_escape}}) | ((?!\\-)[^\]]) )
        (-)
        (?: ({{known_char_escape}}) | ({{invalid_char_escape}}) | ({{char_escape}}) | ([^\]]) )
      captures:
        0: meta.range.regexp
        1: constant.character.escape.regexp
        2: invalid.illegal.character.escape.regexp
        3: constant.character.escape.regexp
        4: constant.character.literal.regexp
        5: punctuation.separator.sequence.regexp-common
        6: constant.character.escape.regexp
        7: invalid.illegal.character.escape.regexp
        8: constant.character.escape.regexp
        9: constant.character.literal.regexp
```

1. While I saw some differences on smaller test files, the performance impract in 5000x lines of following pattern negligible. 

   ```
   [b-c] [\x00-\x{A}] [[a-z]&&[:ascii:]] [^abc\w0-9] [a-c-x] []-] [-a-n] [\-a] [a-z\-9] [\--z] [a-\z] [a-\-]
   ```

2. With regards to maintenance and easier to understand:

   The capture groups are not more complicated than all the numeric constants stuff. It's probably even more straight forward compared to pushing those contexts. The one way or another, I don't see much difference in complexity or readability.

   _Note: `escaped-char` would be obsolete._

3. `with_prototype` impact

   Those capture groups are atomic and can't be broken by injected extra patterns, while those `charset-range-begin` and `charset-range-end` contexts would be prepanded by patterns, which would never match due to the lookahead. This could even slow down the syntax in such scenarios.

   Maybe this can be avoided in future changes, by just not using `with_prototype` though.

The downside of this approach is, it wasn't possible to add further detailed scopes/highlighting to those `known_char_escape` tokens without bloating the capture groups.

Using branch points would be an elegant way, but requires context switches for each character or escape sequence. The prototype I've created performed 15% slower on those lines above.

Reply:
After thinking about it some more, I realised that maybe I *do* have stronger opinions than I thought. I really like that you are asking these questions btw!

Optimising for speed makes perfect sense. Us maintainers don't look at the syntaxes *that* often, we can deal with the added complexity, and saving a few CPU cycles could cumulate across all the users of ST. But is it my preferred option? Honestly, no. One thing I really like is that each context has its own purpose, and should be re-used where possible. Something not highlighting correctly? Only one place to look. New version of the language to support? Easy to find where to make a change. Making a change in one place and having to (remember to) apply it in other places is not my idea of fun (whether it be a scope, a regex etc. I guess for me, its not about capture groups being harder to understand (PackageDev makes it easy when no variables are involved!) but the same construct being scoped in multiple different places) :D I even [had a go at removing the duplication in `group-start` and `group-start-extended`](https://github.com/forkeith/Packages/tree/regex_common_remove_duplicate_group-start-contexts) because I don't like it, and even though the result is a horrible mess of popping multiple contexts at once, I like that it somehow feels easier to understand. My personal opinion ofc! (alas, I didn't get it working for all cases yet, however.)

Nice to know about performance, I'd be tempted to stick to the current approach knowing those numbers, but you raise good points about `with_prototype` - maybe here is where a variable should be used for easy overriding?

Anyway, it feels like we are trying to get a perfect PR, when it may be easier for others to review smaller PRs ;)

Reply:
I am fine with keeping the current approach. I can't really prefer the one or the other solution. So lets keep going with what we have.